rust/agama-cli/src/auth.rs

```diff
@@ -0,0 +1,246 @@
+        AuthCommands::Logout => logout(),
+        AuthCommands::Show => show(),
+    }
+}
+
+/// Reads stored JWT token and returns it
+pub fn jwt() -> anyhow::Result<String> {
+    if let Some(file) = jwt_file() {
+        if let Ok(token) = read_line_from_file(&file.as_path()) {
+            return Ok(token);
+        }
+    }
+
+    return Err(anyhow::anyhow!("JWT not available"));
+}
+
+/// Stores user provided configuration for login command
+#[derive(Args, Debug)]
+pub struct LoginArgs {
+    #[arg(long, short = 'p')]
+    password: Option<String>,
+    #[arg(long, short = 'f')]
+    file: Option<PathBuf>,
+}
+
+impl LoginArgs {
+    /// Transforms user provided options into internal representation
+    /// See Credentials trait
+    fn proceed(options: LoginArgs) -> Box<dyn Credentials> {
+        match options.password {
+            // explicitly provided user + password
+            Some(p) => Box::new(KnownCredentials { password: p }),
+            _ => match options.file {
+                // try to read user + password from a file
+                Some(f) => Box::new(FileCredentials { path: f }),
+                // last instance - ask user to enter user + password interactively
+                _ => Box::new(MissingCredentials {}),
+            },
+        }
+    }
+}
+
+/// Placeholder for no configuration provided by user
+struct MissingCredentials;
+
+/// Stores whatever is needed for reading credentials from a file
+struct FileCredentials {
+    path: PathBuf,
+}
+
+/// Stores credentials as provided by the user directly
+struct KnownCredentials {
+    password: String,
+}
+
+/// Transforms credentials from user's input into format used internaly
+trait Credentials {
+    fn password(&self) -> io::Result<String>;
+}
+
+impl Credentials for KnownCredentials {
+    fn password(&self) -> io::Result<String> {
+        Ok(self.password.clone())
+    }
+}
+
+impl Credentials for FileCredentials {
+    fn password(&self) -> io::Result<String> {
+        read_line_from_file(&self.path.as_path())
+    }
+}
+
+impl Credentials for MissingCredentials {
+    fn password(&self) -> io::Result<String> {
+        let password = read_credential("Password".to_string())?;
+
+        Ok(password)
+    }
+}
+
+/// Path to file where JWT is stored
+fn jwt_file() -> Option<PathBuf> {
+    Some(home::home_dir()?.join(DEFAULT_JWT_FILE))
+}
+
+/// Reads first line from given file
+fn read_line_from_file(path: &Path) -> io::Result<String> {
+    if !&path.exists() {
+        return Err(io::Error::new(
+            io::ErrorKind::Other,
+            "Cannot find the file containing the credentials.",
+        ));
+    }
+
+    if let Ok(file) = File::open(&path) {
+        // cares only of first line, take everything. No comments
+        // or something like that supported
+        let raw = BufReader::new(file).lines().next();
+
+        if let Some(line) = raw {
+            return Ok(line?);
```

Comment:
```suggestion
            return Ok(line);
```

Reply:
No, this won't work because ```BufReader::lines::next``` returns ```Option<Result<...>>``` si line is ```Result``` at this place.

So, it should either stay as is (with ```?```) or ```return line;``` At time of writing I picked the first option to make it kind of obvious that reading of line could fail.

Reply:
Yes, sorry, I made a mistake. I meant `return line`.

Reply:
Actually, `cargo clippy` suggests this change too.